Add state machine config

garimellagowtham opened the pull request:
Added basic interface for creating and storing state configs in state machines. This is work in progress and is not meant to be merged. I am just gathering suggestions to see if things are in the right direction and find the minimal subset that can be merged without interruptions to other people's code.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jhu-asco/aerial_autonomy/124)
<!-- Reviewable:end -->

msheckells commented:
Reviewed 35 of 35 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[include/aerial_autonomy/common/configuration_interface.h, line 12 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwgnqz3hq0AViDBhIFX:-Kwgnqz4kPtUOPB84meA:bl6xpr7) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L12)):*
> ```objc
>  * @brief Dummy base class for all the proto configs.
>  */
> class CommonBase {};
> ```

We should probably call this something like ProtoBase so that we are reminded that it should only be used as a base class for protos.  I wish there was a way to better enforce such a thing...

---

*[include/aerial_autonomy/common/configuration_interface.h, line 21 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwgo50-LUcOMetMWOjp:-Kwgo50-LUcOMetMWOjq:b-q5snn3) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L21)):*
> ```objc
>  * @tparam T type of the object being stored
>  */
> template <class T> class CommonBaseWrapper : public CommonBase {
> ```

Similarly to above, we should class this ProtoWrapper

---

*[include/aerial_autonomy/common/configuration_interface.h, line 48 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgmXt1m3YD54kI5gS7:-KwgmXt1m3YD54kI5gS8:b-dr1xon) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L48)):*
> ```objc
> 
> /**
>  * @brief Class that stores different configuration objects in a typemap
> ```

This class does not have to be specific to storing configurations.  It is pretty general (similarly to TypeMap), so the naming for the class/functions do not need to be specific to configurations.

---

*[include/aerial_autonomy/common/configuration_interface.h, line 60 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwglaB9paxyZ9BjcDc2:-KwglaB9paxyZ9BjcDc3:b-6cg7hn) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L60)):*
> ```objc
>    * Stores a map of the config based on the type index of state. The configs
>    * are
>    * encoded as a tuple of the type index of proto config and encoded string
> ```

not encoded string

---

*[include/aerial_autonomy/common/configuration_interface.h, line 64 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgmBaaR1hvsgIf2IKB:-KwgmBaaR1hvsgIf2IKC:b-poitkk) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L64)):*
> ```objc
>    */
>   std::unordered_map<std::type_index,
>                      std::tuple<const std::type_info *, CommonBase *>>
> ```

why does type_info need to be a pointer?

---

*[include/aerial_autonomy/common/configuration_interface.h, line 78 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwglvZHlLx-ureFSMZs:-KwglvZHlLx-ureFSMZt:b-wjw2cb) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L78)):*
> ```objc
>   template <class TargetState, class ProtoConfig>
>   void addConfig(ProtoConfig config) {
>     auto config_wrapper = new CommonBaseWrapper<ProtoConfig>(config);
> ```

should avoid using a pointer here unless there is a good reason for it

---

*[include/aerial_autonomy/common/thread_safe_state_machine.h, line 53 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwgocwj2nBZ44JDEKiT:-Kwgocwj2nBZ44JDEKiU:be9jy0z) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/thread_safe_state_machine.h#L53)):*
> ```objc
> 
> #define MSM_CONSTRUCTOR_HELPER_EXECUTE_SUB(z, n, unused) ARG##n t##n
> #define MSM_CONSTRUCTOR_HELPER_EXECUTE(z, n, unused)                           \
> ```

What is all this?

---

*[include/aerial_autonomy/state_machines/visual_servoing_state_machine.h, line 101 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwgp0xb-DXIZUjNkTFq:-Kwgp0xcRh6NvR14xlQn:b-upf685) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/state_machines/visual_servoing_state_machine.h#L101)):*
> ```objc
>   VisualServoingStateMachineFrontEnd(
>       UAVVisionSystem &uav_system,
>       const BaseStateMachineConfig &state_machine_config)
> ```

Is there a reason to take a BaeStateMachineConfig as opposed to just taking a specific VisualServoingStateMachineConfig? 

---

*[src/common/system_handler_node_utils.cpp, line 33 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgpqIJlLWTodlz3wzA:-KwgpqIK0SpQfiVmcy5R:b-ocv53h) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/src/common/system_handler_node_utils.cpp#L33)):*
> ```cpp
> 
> UAVSystemHandlerConfig createUAVSystemHandlerConfig(ros::NodeHandle &nh) {
>   std::string uav_system_config_filename;
> ```

There is duplicated code between this function and the above one.  They could be combined into one function which is templated on the config type and which takes the ros param name as an argument.

---

*[tests/common/configuration_interface_tests.cpp, line 49 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgsVRgeXHwXlrsFJlh:-KwgsVRhUT1GRMt8sboK:b1x7k1y) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/tests/common/configuration_interface_tests.cpp#L49)):*
> ```cpp
>   // Try changing the retrieved one
>   retrieved_config1.i = 2;
>   EXPECT_NE(retrieved_config1.i, config1.i);
> ```

shouldn't they be equal since the retrieved config is a reference?

---

*[tests/common/configuration_interface_tests.cpp, line 53 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgsoT7pwJdN9IVvsF-:-KwgsoT7pwJdN9IVvsF0:b-k7et6i) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/tests/common/configuration_interface_tests.cpp#L53)):*
> ```cpp
>   auto retrieved_config_copy =
>       configuration_interface.getConfig<Config1, TargetState1>();
>   EXPECT_EQ(retrieved_config1.i, retrieved_config_copy.i);
> ```

check config for TargetState2 as well

---

*[tests/common/configuration_interface_tests.cpp, line 60 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgseuZ-Yo17oMUTXN5:-KwgseuZ-Yo17oMUTXN6:b1trsqe) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/tests/common/configuration_interface_tests.cpp#L60)):*
> ```cpp
>                std::runtime_error);
> }
> ///
> ```

Add test for overwriting config for a state

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124)*
<!-- Sent from Reviewable.io -->

garimellagowtham commented:
Review status: 28 of 35 files reviewed at latest revision, 12 unresolved discussions.

---

*[include/aerial_autonomy/common/configuration_interface.h, line 12 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwgnqz3hq0AViDBhIFX:-Kwj_UWuq6QRDZ2uR8kq:bj94389) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L12)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

We should probably call this something like ProtoBase so that we are reminded that it should only be used as a base class for protos.  I wish there was a way to better enforce such a thing...
</blockquote></details>

Done. Maybe a better way to enforce is to declare this class inside the `ConfigurationInterface` class.

---

*[include/aerial_autonomy/common/configuration_interface.h, line 21 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwgo50-LUcOMetMWOjp:-Kwj_er0PFd7tdjF6KPh:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L21)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Similarly to above, we should class this ProtoWrapper
</blockquote></details>

Done.

---

*[include/aerial_autonomy/common/configuration_interface.h, line 48 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgmXt1m3YD54kI5gS7:-Kwj_nd0ejQzSDonbwLM:b-atd88d) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L48)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

This class does not have to be specific to storing configurations.  It is pretty general (similarly to TypeMap), so the naming for the class/functions do not need to be specific to configurations.
</blockquote></details>

Should we call this dictionary similar to python? The keys in this dictionary are fixed to be some class types (state types for our state machine use).

---

*[include/aerial_autonomy/common/configuration_interface.h, line 60 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwglaB9paxyZ9BjcDc2:-KwjaXGF--Nx5JlrsaX_:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L60)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

not encoded string
</blockquote></details>

Done.

---

*[include/aerial_autonomy/common/configuration_interface.h, line 64 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgmBaaR1hvsgIf2IKB:-KwjaXeZ9yhMhcAuP2No:bic7hx7) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L64)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

why does type_info need to be a pointer?
</blockquote></details>

Because type_info cannot be copied nor have a default constructor. I also tried type_index and it is also having issues with compilation.

---

*[include/aerial_autonomy/common/configuration_interface.h, line 78 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwglvZHlLx-ureFSMZs:-KwjvMONEviac4XQlsUT:bj0ukje) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L78)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

should avoid using a pointer here unless there is a good reason for it
</blockquote></details>

If I use a temporary variable and pass its address, the config_wrapper will be destroyed after this function.
If I do not use a pointer at all, the config_wrapper is cast as the base class. I will not be able to downcast it correctly anymore. If I remember correctly, this is how you up-cast to a common base class without losing information. I can use a unique ptr or something if that helps!

---

*[include/aerial_autonomy/common/thread_safe_state_machine.h, line 53 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwgocwj2nBZ44JDEKiT:-Kwjb-fm2qxIbRsd7xx2:b7t1umw) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/thread_safe_state_machine.h#L53)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

What is all this?
</blockquote></details>

This defines new constructors for state machine with custom arguments. Initially, we had only one argument. But, with adding a state machine config, we need this stuff to define the corresponding constructor for back-end.

---

*[include/aerial_autonomy/state_machines/visual_servoing_state_machine.h, line 101 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwgp0xb-DXIZUjNkTFq:-KwjbMp771fPb5H3BeDQ:b-1y74hs) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/state_machines/visual_servoing_state_machine.h#L101)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Is there a reason to take a BaeStateMachineConfig as opposed to just taking a specific VisualServoingStateMachineConfig? 
</blockquote></details>

Currently, the state machine is created by the common system handler class. This requires all the state machine to have some common arguments. We can add another template argument (The proto config) to common system handler and that way we can have different configurations for different common system handlers.

---

*[src/common/system_handler_node_utils.cpp, line 33 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgpqIJlLWTodlz3wzA:-Kwjz1TZxtL1onOXoaW1:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/src/common/system_handler_node_utils.cpp#L33)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

There is duplicated code between this function and the above one.  They could be combined into one function which is templated on the config type and which takes the ros param name as an argument.
</blockquote></details>

Done.

---

*[tests/common/configuration_interface_tests.cpp, line 49 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgsVRgeXHwXlrsFJlh:-Kwjz2Esq9EznO2zlOMF:bndohht) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/tests/common/configuration_interface_tests.cpp#L49)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

shouldn't they be equal since the retrieved config is a reference?
</blockquote></details>

The addConfig is only copying the object. So the reference is only corresponding to the one stored in the configuration interface. [Discuss in person]

---

*[tests/common/configuration_interface_tests.cpp, line 53 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgsoT7pwJdN9IVvsF-:-Kwk3nm9gW_5wgsUJUvL:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/tests/common/configuration_interface_tests.cpp#L53)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

check config for TargetState2 as well
</blockquote></details>

Done.

---

*[tests/common/configuration_interface_tests.cpp, line 60 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgseuZ-Yo17oMUTXN5:-Kwk3kVE3tgKFvzPVvIj:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/tests/common/configuration_interface_tests.cpp#L60)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Add test for overwriting config for a state
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124)*
<!-- Sent from Reviewable.io -->

garimellagowtham commented:
Review status: 27 of 39 files reviewed at latest revision, 12 unresolved discussions.

---

*[include/aerial_autonomy/common/configuration_interface.h, line 12 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwgnqz3hq0AViDBhIFX:-Kwkgk-d_RWsGhlFmSeN:b-miw4as) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L12)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Done. Maybe a better way to enforce is to declare this class inside the `ConfigurationInterface` class.
</blockquote></details>

Moved ProtoWrapper into configuration interface.

---

*[include/aerial_autonomy/common/configuration_interface.h, line 48 at r1](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwgmXt1m3YD54kI5gS7:-Kwl2KN4w0swLtxTd7uO:buzarae) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/14a9dacc613bacf05017d9b85cc8d99e9ae9042f/include/aerial_autonomy/common/configuration_interface.h#L48)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Should we call this dictionary similar to python? The keys in this dictionary are fixed to be some class types (state types for our state machine use).
</blockquote></details>

Renamed to `UnorderedHeterogenousMap`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124)*
<!-- Sent from Reviewable.io -->

msheckells commented:
Reviewed 14 of 14 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 21 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu4mZByAMV9SK9NNYp:-Kwu4mZByAMV9SK9NNYq:b-33f72x) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L21)):*
> ```objc
>    * @brief Dummy base class for all the value objects
>    */
>   class AbstractBase {};
> ```

Not really abstract because it does not have any pure virtual functions... Maybe just call it Base?

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 56 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu51c04P-x3rzSzWmZ:-Kwu51c04P-x3rzSzWm_:b-6t51cx) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L56)):*
> ```objc
>   };
>   /**
>    * Stores a map of the config based on the type index of state. The configs
> ```

Stores a map from Key to Base objects

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 68 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu5EUtjAmgKWWvmVTV:-Kwu5EUtjAmgKWWvmVTW:b-r2duu) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L68)):*
> ```objc
> public:
>   /**
>    * @brief add a configuration object into typemap
> ```

Add Key/Value pair to map

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 74 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu5XzcKWTRPAhLTm4a:-Kwu5XzcKWTRPAhLTm4b:b-6leyt9) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L74)):*
> ```objc
>    * @param value The configuration object to be stored
>    */
>   template <class Value> void addConfig(const Key &key, Value value) {
> ```

Call this "insert" just to be consistent with other C++ containers

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 81 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu5jZ_f2HaXx6RpO31:-Kwu5jZ_f2HaXx6RpO32:bctnujc) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L81)):*
> ```objc
>   }
>   /**
>    * @brief Get the configuration object stored in typemap
> ```

Get the value for a given Key.  Throws std::runtime_error if Key is not present

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 88 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu5ehB81y48lhzSB1n:-Kwu5ehB81y48lhzSB1o:b-ymrino) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L88)):*
> ```objc
>    * @return A copy of the configuration object stored in typemap
>    */
>   template <class Value> Value getConfig(const Key &key) const {
> ```

Call this "find"

---

*[tests/common/log_config.pbtxt, line 1 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu6CxhqYhZAi2Isb-L:-Kwu6CxhqYhZAi2Isb-M:b-r7wo5o) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/log_config.pbtxt#L1)):*
> ```
> # This file is automatically generated! Make changes to log_config.pbtxt.in
> ```

This file actually should not be in git since it is automatically generated. Can you add it to .gitignore?

---

*[tests/common/unordered_heterogenous_map_tests.cpp, line 28 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu72w2RtgRvyKq8rot:-Kwu72w3tYIuqaOYQhCw:b-43nc09) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/unordered_heterogenous_map_tests.cpp#L28)):*
> ```cpp
>       config1_wrapper(original);
>   // Retrieve object
>   retrieved = config1_wrapper.getInput();
> ```

This is just testing if the wrapper works? Should change the test name if so.

---

*[tests/common/unordered_heterogenous_map_tests.cpp, line 76 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwuPpud9Vbm1ZkFi38L:-KwuPpud9Vbm1ZkFi38M:b-5kjhga) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/unordered_heterogenous_map_tests.cpp#L76)):*
> ```cpp
>   config1.i = 10;
>   Config1 config1_overwrite;
>   config1.i = 11;
> ```

is this supposed to be config1_overwrite?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124)*
<!-- Sent from Reviewable.io -->

msheckells commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[tests/common/unordered_heterogenous_map_tests.cpp, line 33 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwuQ0z0heKxRiiPZSLQ:-KwuQ0z1KvAE4LlT13Jb:b-4bygje) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/unordered_heterogenous_map_tests.cpp#L33)):*
> ```cpp
>   Config1 &retrieved_ref = config1_wrapper.getInput();
>   retrieved_ref.i = 10;
>   retrieved = config1_wrapper.getInput();
> ```

remove commented code

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124)*
<!-- Sent from Reviewable.io -->

msheckells commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[tests/common/unordered_heterogenous_map_tests.cpp, line 45 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwuQNEVQVorAPlthi4N:-KwuQNEVQVorAPlthi4O:bmsuq4i) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/unordered_heterogenous_map_tests.cpp#L45)):*
> ```cpp
>   config2.j = 5;
>   // Add configs to interface
>   configuration_interface.addConfig(TYPE_INDEX(TargetState1), config1);
> ```

Using type_index explicitly is fine for now.  In the future, we should maybe have a specialized UnorderedHeterogenousMap whose Key is a type_index and whose insert function is templated on the type to be used as the key, e.g. `map.insert<TargetState1>(config1)`.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124)*
<!-- Sent from Reviewable.io -->

garimellagowtham commented:
Review status: 36 of 39 files reviewed at latest revision, 11 unresolved discussions.

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 21 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu4mZByAMV9SK9NNYp:-KwuUCGQ1H607BsarLQl:b-iaw75v) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L21)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Not really abstract because it does not have any pure virtual functions... Maybe just call it Base?
</blockquote></details>

Converted it into an abstract class by making the destructor pure virtual with a default implementation. So you cannot instantiate abstract: 
https://stackoverflow.com/questions/4640985/making-a-class-abstract-without-any-pure-virtual-methods

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 56 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu51c04P-x3rzSzWmZ:-KwuXYUpPN8Psu-GkJyM:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L56)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Stores a map from Key to Base objects
</blockquote></details>

Done.

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 68 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu5EUtjAmgKWWvmVTV:-KwuXegnzNUg-z_ANj7r:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L68)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Add Key/Value pair to map
</blockquote></details>

Done.

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 74 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu5XzcKWTRPAhLTm4a:-KwuXk2JWcn2vR3kSr3y:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L74)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Call this "insert" just to be consistent with other C++ containers
</blockquote></details>

Done.

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 81 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu5jZ_f2HaXx6RpO31:-KwuYGTbzKAWhA9Ywa1n:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L81)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Get the value for a given Key.  Throws std::runtime_error if Key is not present
</blockquote></details>

Done.

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 88 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu5ehB81y48lhzSB1n:-KwuY_LAYx3A-I97HAG6:bb6un1j) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L88)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Call this "find"
</blockquote></details>

Done

---

*[tests/common/log_config.pbtxt, line 1 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu6CxhqYhZAi2Isb-L:-KwuYcEgj4co9bdx2k34:b-f9ksg8) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/log_config.pbtxt#L1)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

This file actually should not be in git since it is automatically generated. Can you add it to .gitignore?
</blockquote></details>

This is not generated automatically. Removed comment. The auto-generated one is in params folder. This file is only for testing the log config stuff.

---

*[tests/common/unordered_heterogenous_map_tests.cpp, line 28 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu72w2RtgRvyKq8rot:-KwuqDPvOK37woSjpQ-D:bfv6ldg) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/unordered_heterogenous_map_tests.cpp#L28)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

This is just testing if the wrapper works? Should change the test name if so.
</blockquote></details>

It is saving and retrieving an object from the wrapper.

---

*[tests/common/unordered_heterogenous_map_tests.cpp, line 33 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwuQ0z0heKxRiiPZSLQ:-KwuqcDgRhnPzixpmaig:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/unordered_heterogenous_map_tests.cpp#L33)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

remove commented code
</blockquote></details>

Done.

---

*[tests/common/unordered_heterogenous_map_tests.cpp, line 45 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwuQNEVQVorAPlthi4N:-Kwv4tWAHarT4ghrMypv:b-w6bedq) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/unordered_heterogenous_map_tests.cpp#L45)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

Using type_index explicitly is fine for now.  In the future, we should maybe have a specialized UnorderedHeterogenousMap whose Key is a type_index and whose insert function is templated on the type to be used as the key, e.g. `map.insert<TargetState1>(config1)`.  
</blockquote></details>

Implemented two functions with the specified interface.

---

*[tests/common/unordered_heterogenous_map_tests.cpp, line 76 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-KwuPpud9Vbm1ZkFi38L:-Kwv4xLAk0KUmnbMUs9t:b-896fix) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/tests/common/unordered_heterogenous_map_tests.cpp#L76)):*
<details><summary><i>Previously, msheckells (Matt Sheckells) wrote…</i></summary><blockquote>

is this supposed to be config1_overwrite?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124)*
<!-- Sent from Reviewable.io -->

msheckells commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[include/aerial_autonomy/common/unordered_heterogenous_map.h, line 21 at r2](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124#-Kwu4mZByAMV9SK9NNYp:-Kx4tAPSQukkoe3Eb7i5:bb6u032) ([raw file](https://github.com/jhu-asco/aerial_autonomy/blob/696f1c59dd1bd23608d88c18cc2df8fa594adec1/include/aerial_autonomy/common/unordered_heterogenous_map.h#L21)):*
<details><summary><i>Previously, garimellagowtham (Gowtham Garimella) wrote…</i></summary><blockquote>

Converted it into an abstract class by making the destructor pure virtual with a default implementation. So you cannot instantiate abstract: 
https://stackoverflow.com/questions/4640985/making-a-class-abstract-without-any-pure-virtual-methods
</blockquote></details>

Cool

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jhu-asco/aerial_autonomy/124)*
<!-- Sent from Reviewable.io -->
